### PR TITLE
CANDLEPIN-323: Update exception handling for adapters [M]

### PR DESCRIPTION
**Pull request by wottop**
- Brings the translated messages to Candlepin away from adapter implementation code.
- Exceptions from the adapters are qualified so that Candlepin can react to specific types.

### CODE REVIEW

**Comment by wottop** [comment]
This will need to stay in draft until the accompanying changes are made in the service adapters. The specific exception changes in the adapters will need to wait until the adapters are first updated to CP 4.3.6.

Feedback on this branch in the meantime is appreciated.

**Review by Januson — CHANGES_REQUESTED** [review]
Looks good. There are a few things to iron out before it's ready. Also two additional thoughts on the overall change.

- How thorough we want to be with testing? We should probably unit test all these mappings. 
- It might be better to use an exception mapper instead of mapping in place of occurrence. E.G. UserServiceExceptionMapper that would handle transformation of all UserServiceExceptions (See RuntimeExceptionMapper for an example). But that would probably require having those exceptions unchecked to avoid polluting the code with redundant throws.

File: src/main/java/org/candlepin/resource/CloudRegistrationResource.java
Comment (Januson) [inline src/main/java/org/candlepin/resource/CloudRegistrationResource.java:83]: One more thought. Rather than catching and mapping, we could provide hosted with a set of exceptions that have the translation integrated inside. That way we wouldn't need to do any mapping on our side. All hosted exceptions would be translated. E.G.

```java
public class CloudRegistrationBadTypeException extends ProductServiceException {
    public CloudRegistrationBadTypeException(I18n i18n, String type) {
        super(i18n.tr("Unrecognized provider type in request: {}", type));
    }
}
```

File: src/main/java/org/candlepin/service/exception/user/UserMissingOwnerException.java
Comment (redsaz) [inline src/main/java/org/candlepin/service/exception/user/UserMissingOwnerException.java:21]: Adapters had (well, still has) been using an OwnerCurator in order to look up the full details of the Owner(s) that the user is a part of, as part of the user lookup. When the OwnerCurator returns back null for "nah, there's no owner by that key". then Adapters would either throw an error or return back an empty list of accessible owners depending on the method.

I posed a question in a google meet call with Candlepin folks, including @Ceiu about Adapters use of curators. It was determined that _*any use of curators by Adapters*_ is ill-informed, no good, bad, why-is-this-happening-ow-my-eyes.

So, at some point down the line, and hopefully soon, Adapters will no longer be using the owner curator to fill in the owner details when looking up a user, and will instead return an `OwnerInfo` (though if an `OwnerKeyInfo` was a thing I'd do that) which only has info for `key`. Which means that Adapters won't be erroring on missing owners, and it'll now be on core Candlepin to determine what to do when it gets back a user who belongs to an Owner that doesn't exist. Which means that this exception probably isn't needed, but not 100% sure on that. At the very least, eventually Adapters will never throw the exception, because in order to throw it, Adapters must be able to look up the Owner, and the only way it can do that is through Curator, which Adapters should not use.

File: src/main/java/org/candlepin/resource/OwnerResource.java
Comment (redsaz) [inline src/main/java/org/candlepin/resource/OwnerResource.java:1290]: `OwnerInvalidKeyFormatException` looks like it is intended for use by this Adapter code:

```java
  public boolean isOwnerKeyValidForCreation(String ownerKey) {
    try {
      LOGGER.debug("Checking if ownerkey={} is valid.", ownerKey);
      return userService.isOwnerKeyValid(Long.parseLong(ownerKey));
    } catch (NumberFormatException e) {
      LOGGER.debug("... ownerkey={} is invalid due to NumberFormatException.", ownerKey);
      throw new CandlepinAdapterException(ownerKey + " is not a valid OrgId", e);
    }
  }
```

I don't believe I ever was happy about the exception getting thrown by Adapters here. To me, it should have simply returned `false`, because that is the answer to the `isOwnerKeyValidForCreation` question.

With that in mind, are you okay if we get rid of OwnerInvalidKeyFormatException, and Adapters can not throw an exception here, and return `false` instead?

File: src/main/java/org/candlepin/service/exception/subscription/SubscriptionActivationNoTagException.java
Comment (redsaz) [inline src/main/java/org/candlepin/service/exception/subscription/SubscriptionActivationNoTagException.java:20]: Looks like a minor auto-formatter issue here, extra asterisk

File: src/main/java/org/candlepin/service/exception/subscription/SubscriptionSuccessfulActivationException.java
Comment (redsaz) [inline src/main/java/org/candlepin/service/exception/subscription/SubscriptionSuccessfulActivationException.java:22]: I did a double-take here and asked "Wait, do the Adapters actually do this?" And sure enough, yes it does.

In fact, the `activateSubscription` method never returns, it always throws an exception. In this case, a successful exception. Which seems wacky to me.

We'll have more opportunities to improve this adapter later, I think we can keep all the exceptions. What are your thoughts on putting all these subscription activation exceptions in a sub package, say `org.candlepin.exception.subscription.activation` to keep them separate from the more traditional subscription exceptions?

(Another aside, what are thoughts on extracting the activation methods and exceptions out of the `SubscriptionServiceAdapter` and into a `SubscriptionActivationServiceAdapter`? This would be effort for a different, future ticket, not this one.)

File: src/main/java/org/candlepin/service/exception/subscription/SubscriptionUnknownActivationException.java
Comment (redsaz) [inline src/main/java/org/candlepin/service/exception/subscription/SubscriptionUnknownActivationException.java:22]: This looks like another potential candidate to be replaced by a suggestion in an earlier comment, to have a generic `ServiceRuntimeException` instead. Then, in the SubscriptionResource block:

```java
           else if (e instanceof SubscriptionUnknownActivationException) {
                message = i18n.tr(
                    "The system is unable to redeem the requested subscription: {0}", dellAssetTag);
            }
```

Instead of catching `SubscriptionUnknownActivationException`. it can catch the `ServiceRuntimeException` and still provide the same error message.

File: src/main/java/org/candlepin/service/exception/subscription/SubscriptionUnknownRetrievalException.java
Comment (redsaz) [inline src/main/java/org/candlepin/service/exception/subscription/SubscriptionUnknownRetrievalException.java:22]: Parroting earlier comments, this could be replaced by a potential `ServiceRuntimeException` or `UnknownServiceException` general case exception, and the catch block that receives the exceptions can still provide the correct logging for the case.

File: src/main/java/org/candlepin/service/exception/subscription/SubscriptionUnknownTermsException.java
Comment (redsaz) [inline src/main/java/org/candlepin/service/exception/subscription/SubscriptionUnknownTermsException.java:22]: Dittoing earlier comments, this could be replaced by a potential ServiceRuntimeException or UnknownServiceException general case exception, and the catch block that receives the exceptions can still provide the correct logging for the case.

File: src/main/java/org/candlepin/service/exception/user/UserMissingException.java
Comment (redsaz) [inline src/main/java/org/candlepin/service/exception/user/UserMissingException.java:21]: For now, having both a UserInvalidException and UserMissingException is okay. Today, Adapters **does** distinguish between invalid users (users with incorrect login+password combos) and users that do not exist (login was not found) for certain calls. However, I wonder if we should be returning UserInvalidException in either case, for security purposes.

File: src/main/java/org/candlepin/service/exception/user/UserRetrievalException.java
Comment (redsaz) [inline src/main/java/org/candlepin/service/exception/user/UserRetrievalException.java:21]: It doesn't appear that this class gets caught anywhere, could this be deleted?

File: src/main/java/org/candlepin/service/exception/user/UserUnknownRetrievalException.java
Comment (redsaz) [inline src/main/java/org/candlepin/service/exception/user/UserUnknownRetrievalException.java:21]: Dittoing earlier comments, this could be replaced by a potential `ServiceRuntimeException` or `UnknownServiceException` general case exception, and the catch block that receives the exceptions can still provide the correct logging for the case.

**Review by redsaz — CHANGES_REQUESTED** [review]
This does a good job covering all the different cases that Adapters was covering with its own custom exceptions and i18n translations. There were a couple of exceptions that might've let too much implementation details leak through, which I commented on. There were a few different `*UnknownException` exceptions that I thought could merge into a single `ServiceRuntimeException` since they would cover transient errors that are outside of Candlepin's control.

Out of curiosity, where did the i18n translations go? I don't see any .po files get changed/updated within this commit.

File: src/main/java/org/candlepin/resource/OwnerResource.java
Comment (Ceiu) [reply src/main/java/org/candlepin/resource/OwnerResource.java:1290]: That seems reasonable to me

File: src/main/java/org/candlepin/service/exception/product/ProductUnknownRetrievalException.java
Comment (Ceiu) [reply src/main/java/org/candlepin/service/exception/product/ProductUnknownRetrievalException.java:22]: I would agree with this. All of our service exceptions should derive from a common exception. Maybe a pair of common exceptions for both checked and unchecked variations: ServiceException (checked) and ServiceRuntimeException (unchecked).

File: src/main/java/org/candlepin/service/exception/user/UserMissingException.java
Comment (Ceiu) [reply src/main/java/org/candlepin/service/exception/user/UserMissingException.java:21]: This is an interesting one. I guess the question comes down to whether or not Candlepin ever needs to care about such a differentiation and act accordingly. If not, then the adapters throwing a single "user not found" exception is probably okay. Otherwise, it may make sense to leave that up to CP to kick back the same exception in both paths.

But to the larger point, I agree that we could be more limiting on the info we give to users via error messages.

**Comment by nikosmoum** [comment]
The new exceptions added by Stepan for autoreg 2.0 potentially need an update here to adhere to the pattern we chose

File: src/main/java/org/candlepin/service/exception/product/ProductUnknownRetrievalException.java
Comment (wottop) [reply src/main/java/org/candlepin/service/exception/product/ProductUnknownRetrievalException.java:22]: I did want to keep the "UnknownServiceException" at least be categorized as to the service that threw it. Since the  end goal is to stop passing messages through exceptions, then this differentiation is useful.

**Comment by wottop** [comment]
The .po files are updated just before a tagged release.

File: src/main/java/org/candlepin/service/exception/subscription/SubscriptionSuccessfulActivationException.java
Comment (wottop) [reply src/main/java/org/candlepin/service/exception/subscription/SubscriptionSuccessfulActivationException.java:22]: Updated both ends such that there is no exception returned on a successful action.

File: src/main/java/org/candlepin/auth/BasicAuth.java
Comment (Ceiu) [reply src/main/java/org/candlepin/auth/BasicAuth.java:89]: They should be. Linting takes a backseat to better code structure.

**Comment by wottop** [comment]
This can go to final review now. I talked with Shayne and we feel it should be put in a tag by itself so we can get it to stage with the updated adapter code. If there is a big problem (which we don't expect) then it can be reverted. There is not another way to effectively smoke test it and we should see any issues quickly.

File: src/main/java/org/candlepin/service/exception/product/ProductUnknownRetrievalException.java
Comment (Ceiu) [reply src/main/java/org/candlepin/service/exception/product/ProductUnknownRetrievalException.java:21]: A quick look at the adapter level changes, I drafted up an exception tree that looks something like this:

```
ServiceAdapterException
    CloudRegistrationServiceException
        CloudRegistrationAuthorizationException
        MalformedCloudRegistrationException

    EntitlementCertServiceException

    IdentityCertServiceException

    OwnerServiceException

    ProductServiceException

    SubscriptionServiceException
        ExpiredTagException
        ExhaustedTagException (tag already used)
        InvalidTagException

    UserServiceException
        InvalidCredentialsException
        UnacceptedTermsException
        UnauthorizedException
        UserDisabledException
```

This may not be a complete capture of everything (some of the calls were a bit difficult to follow), but it gives the outline of what the exception tree should look like. The remaining question is whether or not we consider these true error states or actionable and thus whether or not they should be checked or unchecked exceptions. Sadly with a tree like this, we're kinda forced into one or the other barring some shenanigans I'm not aware of.

I'm kinda leaning toward unchecked exceptions, since the only thing we really do with them is translate them to the appropriate error code, and not catching one should probably result in a 500 level as a signal that we need to catch it. In either case we're going to need the try/catch blocks around these calls, so the boilerplate is still going to be a thing in many cases, but I can be convinced either way yet.

**Review by Ceiu — CHANGES_REQUESTED** [review]
If time permits, we should go over this again next week.

File: src/main/java/org/candlepin/controller/Refresher.java
Comment (Ceiu) [inline src/main/java/org/candlepin/controller/Refresher.java:193]: We shouldn't be using CandlepinExceptions at this level, as we're not API-facing. These should either be rethrown directly or wrapped in something like a RuntimeException or something more specific if applicable.

File: src/main/java/org/candlepin/resource/ConsumerResource.java
Comment (Ceiu) [inline src/main/java/org/candlepin/resource/ConsumerResource.java:2337]: Does this need to be handled in a nested try-catch? On the surface it looks like we can avoid this nesting and rearranging of code via finally without it.

File: src/main/java/org/candlepin/resource/ConsumerResource.java
Comment (Ceiu) [inline src/main/java/org/candlepin/resource/ConsumerResource.java:2437]: No apostrophe on "IDs".

File: src/main/java/org/candlepin/resource/OwnerResource.java
Comment (Ceiu) [inline src/main/java/org/candlepin/resource/OwnerResource.java:1294]: Is the line break necessary here?

File: src/main/java/org/candlepin/service/exception/subscription/SubscriptionServiceExceptionMapper.java
Comment (Ceiu) [inline src/main/java/org/candlepin/service/exception/subscription/SubscriptionServiceExceptionMapper.java:21]: While I'm not a huge fan of these mappers, if we're going to have them at all, they should all function the same way. This mapper returns a string for the pre-translated message to be, presumably, wrapped in another exception; rather than just throwing the exception like the other mappers do.

So far this seems to be the odd one out, and given the point of abstraction out the handling of specific exceptions is to avoid boilerplate, this class should follow in the convention of returning (not throwing!) the appropriate exception established by the other mappers.

As a side note, since the mappers are throwing Candlepin or API-facing exceptions, these should only ever be used from the API handlers (typically "resource" classes). Additionally, the mappers shouldn't be forced to accept a bunch of state to build the error messages needed for the mapped exceptions -- that bit should be contained within the exception itself. I would expect the mapper call to be nothing more than:

`throw SubscriptionServiceExceptionMapper.map(exception);`

For the exceptions that wrap additional state, that should be present in their constructors. E.g. `throw new SubscriptionExpiredTagException(tag)`

File: src/main/java/org/candlepin/service/exception/subscription/SubscriptionServiceExceptionMapper.java
Comment (Ceiu) [inline src/main/java/org/candlepin/service/exception/subscription/SubscriptionServiceExceptionMapper.java:29]: Adding onto the comment above, this handler runs directly into the problem these mappers can hit without care. `SubscriptionActivationException` is pretty generic, but because of the state required to be passed in with the mapping function itself, there's an assumption that this only occurs when registering with an asset tag, and that is reflected in the output message. This limits the scope of the mapper to only the one or two places where tag-based activation is invoked, somewhat defeating the point of abstracting this logic out.

**Review by Ceiu — CHANGES_REQUESTED** [review]
This is coming along, but the handling of exceptions still needs a bit of work

File: src/main/java/org/candlepin/resource/CloudRegistrationResource.java
Comment (Ceiu) [inline src/main/java/org/candlepin/resource/CloudRegistrationResource.java:68]: Line break between the end of this block and start of the next

**Review by Ceiu — CHANGES_REQUESTED** [review]
Getting there; just need a bit of refinement on the exception properties (and likely translation).

I imagine one of the reasons it was implemented the way it was is to avoid the tedium of all the conversion with the instanceof spam in the mappers. If that's accurate, we have a few options:

1) Java 17 gives us the instanceof "pattern matching" syntax to avoid the need for explicit conversion:
```
if (exception instanceof UserDisabledException ude) {
  System.out.println("User disabled: " + ude.getUsername());
}
```

2) Something a bit jank we could do as well -- though I'm not sure how to feel about it -- would be to use try-catch to cast the exceptions in the mapper:
```
public CandlepinException map(UserServiceException exception) {
  try {
    throw exception;
  }
  catch (UserDisabledException ude) {
    return new BadRequestException("user disabled: " + ude.getUsername());
  }
  ...
  catch (UserServiceException use) {
    return new BadRequestException("generic error msg");
  }
}
```

I think either method is going to come with some degree of pain and ugliness, since it's honestly a pretty janky way of dealing with exceptions. It really should be done on a case-by-case basis rather than abstracted out; but so long as we're going to be doing it this way, then we'll have to live with the pain that comes along with it.

**Review by Ceiu — CHANGES_REQUESTED** [review]
The core logic changes look to be complete. At this point it's just stylistic and documentation cleanup.

The main comment of my last review is still pending as well -- we could be using newer Java features to avoid needing to explicitly cast the exceptions in the mappers.

Also, we should get another set of eyes on this to chime in on the mappers. For what it's worth, I did a quick count of their use and my results were:
- user service: 7
- cloud registration: 1
- subscription service: 1

Those mappers which are only used once could be removed and have the places using them manually handle the exceptions directly.

**Comment by wottop** [comment]
I will squash down after approval.

File: src/main/java/org/candlepin/service/exception/ServiceException.java
Comment (Ceiu) [inline src/main/java/org/candlepin/service/exception/ServiceException.java:75]: Mutable properties on the exception again. Also what was this intended to be used for? Looks to be extraneous with the mapper.

**Review by Ceiu — CHANGES_REQUESTED** [review]

File: src/main/java/org/candlepin/service/exception/ServiceException.java
Comment (Ceiu) [inline src/main/java/org/candlepin/service/exception/ServiceException.java:79]: Hrmm... if the mapper is reliant on stuff like this, we're better off with the gigabranch that the original had. Definitely would like to see something slightly more involved like a function or consumer so we don't need to add generic abstractions like this to the exception API. Something closer to this:

```
mappings.add(ProductServiceException.class, pse -> {
  return Pair.of(
    Status.BAD_REQUEST,
    i18n.tr("Unable to retrieve product \"{0}\" from ProductService.", pse.getProductId()));
})
```

We can then also use `.getOrDefault` to wrap up the generic message stuff as well:

```
Pair<Status, String> result = this.mappings.getOrDefault(exception.getClass(), 
  e -> Pair.of(Status.BAD_REQUEST, i18n.get().tr("Unknown service exception"));

return this.getDefaultBuilder(..., result.getLeft());
```

File: src/main/java/org/candlepin/exceptions/mappers/ServiceExceptionMapper.java
Comment (Ceiu) [inline src/main/java/org/candlepin/exceptions/mappers/ServiceExceptionMapper.java:118]: Note that this lookup doesn't work with subclasses.

When/if we move to Java 21, we can start using the switch expressions to best wrap all of this up in a single mapping function, but for now we may need to either do a semi-recursive lookup, or at least document that subclasses of registered mappings aren't implicitly handled as one might expect.

@Januson thoughts on this?

Comment (Januson) [reply src/main/java/org/candlepin/exceptions/mappers/ServiceExceptionMapper.java]: I agree. Switch would make it a bit more concise, but until then, pattern matching with instanceOf should be good enough.

**Pull request merged**
